File: battle/Player.py
import Pokemon as p
import json, random

class Player():
	def __init__(self, filepath, team=None, knownOpponentTeam=None, opponentTeam=None):
		if team is None:
			self.team = {}
		if knownOpponentTeam is None:
			self.knownOpponentTeam = {}
		if opponentTeam is None:
			self.opponentTeam = {}
		self.has_won = False;

		self.build_team(filepath)

	def set_current_out(self, move):
		if move.isSwitch:
			self.current_out = move.pokemon.name

	def set_opponent_current_out(self, move):
		if move.isSwitch:
			self.opponent_current_out = move.pokemon.name

	""" This method handles the logic of updating the opponent's team
	i.e. changing which Pokemon is currently out, which Pokemon are known,
	and what moves are known.

	After this method is called, everything except damage will be correctly updated
	in the player's model of the opponent's team. """
	def update_opponent_team(self, move):
		if move.isSwitch:
			if move.pokemon.name not in self.knownOpponentTeam:
				newPokemon = p.Pokemon(move.pokemon.name)
				self.knownOpponentTeam[move.pokemon.name] = newPokemon
				self.set_opponent_current_out(move)
			else:
				self.set_opponent_current_out(move)
		else:
			pokemon = self.knownOpponentTeam[move.pokemon.name]
			if move.moveName not in pokemon.moves:
				pokemon.moves.append(move.moveName)

	def process_turn(self, your_move, opponent_move):
		self.update_opponent_team(opponent_move)
		self.set_current_out(your_move)


	def move(self):
		move_type = random.randint(0,1)
		if move_type is 0:
			switch = self.current_out
			while (switch == self.current_out):
				switch = random.choice(list(self.team.keys()))
			move = player_move(True, self.team[switch])
			print("Switched to " + switch)
		else:
			i = random.randint(0,3)
			move = player_move(False, self.team[self.current_out], self.team[self.current_out].moves[i])
			print(self.current_out + " used " + self.team[self.current_out].moves[i])

		return move

	"""
	The format of each individual pokemon is this:
	Jynx
	Blizzard
	Psychic
	Lovely Kiss
	Rest
	newline
	...
	"""
	def build_team(self, filepath):
		with open(filepath, "r") as infile:
			lines = infile.readlines()

		for i in range(len(lines)):
			if i % 6 == 0:
				name = lines[i].strip("\n").replace(" ","-")
				newPokemon = p.Pokemon(name)
				self.team[name] = newPokemon
			elif not (i % 6 == 5):
				self.team[name].addMove(lines[i].replace(" ","").lower().strip("\n"))
			elif i == 5:
				self.set_current_out(player_move(True, newPokemon))


	def __str__(self):
		print("i exist!!")

class player_move():
	def __init__(self, isSwitch, pokemon, moveName=None):
		self.isSwitch = isSwitch
		self.pokemon = pokemon
		self.moveName = moveName if moveName is not None else ""

if __name__ == "__main__":
	thomas = Player("sample-team.txt")
	for i in range(10):
		move = thomas.move()
		thomas.process_turn(move, player_move(True, p.Pokemon("Tauros")))

File: JuniorDesign_CLion/Player.cpp
#include "Player.h"


Player::Player(std::string filepath,
	 std::unordered_map<std::string, Pokemon> team,
	 std::unordered_map<std::string, Pokemon> knownOpponentTeam,
	 std::unordered_map<std::string, Pokemon> opponentTeam)
{
	srand(time(0)); //TODO: there is almost definitely a better place for this
	this->team = team; //TODO: We're setting the team here and in buildTeam, is this necessary?
	this->knownOpponentTeam = knownOpponentTeam;
	this->opponentTeam = opponentTeam;
	this->hasWon = false;
	buildTeam(filepath);
}

Player::Player(string filepath)
{
	this->hasWon = false;
	buildTeam(filepath);
}

void Player::setCurrentOut(PlayerMove move)
{
	if (move.isSwitch)
		this->currentOut = move.pokemon.getName();
}

void Player::setOpponentCurrentOut(PlayerMove move)
{
	if (move.isSwitch)
		this->opponentCurrentOut = move.pokemon.getName();
}

/*	This method handles the logic of updating the opponent's team
 *  i.e. changing which Pokemon is currently out, which Pokemon are known,
 *  and what moves are known.
 
 *	After this method is called, everything except damage will be correctly updated
 *  in the player's model of the opponent's team.
 */
void Player::updateOpponentTeam(PlayerMove move)
{
	if (move.isSwitch)
	{
		if (this->knownOpponentTeam.find(tove.pokemon.getName()) == this->knownOpponentTeam().end())
		{
			Pokemon *p = new Pokemon(move.pokemon.getName());
			this->knownOpponentTeam[move.pokemon.getName()] = *p;
			setOpponentCurrentOut(move);
		} else
		{
			setOpponentCurrentOut(move);
		}
	} else 
	{
		Pokemon *p = new Pokemon(this->knownOpponentTeam(move.pokemon.getName()));
		if (!p->containsMove(move.moveName))
			pokemon.appenMove(move.moveName);
	}
}

PlayerMove Player::move()
{
	int moveType = rand() % 2;
	PlayerMove move;
	if (!moveType)
	{
		string switchPokemon = this->currentOut;
		while (switchPokemon == this->currentOut)
		{
			int randomChoice = rand() % this->team.size();
			auto it = this->team.begin();
			for (int i = 0; i < randomChoice; i++)
			{
				switchPokemon = it->first;
				it++;
			}
		}
		move.isSwitch = true;
		move.Pokemon = &this->team[switchPokemon];
		std::cout << "Switched to " << switchPokemon << std::endl;
	} else 
	{
		int i = rand() % 4;
		move.isSwitch = false;
		move.Pokemon = &this->team[this->currentOut];
		move.moveName = this->team[this->currentOut].getMove(i).moveName;
		std::cout << this->currentOut 
			<< " used " 
			<< this->team[this-currentOut].getMove(i).moveName
			<< std::endl;
	}
	return move;
}

void Player::buildTeam(std::string filepath)
{
	std::ifstream infile(filepath);
	std::string line;
	int i = 0;
	while (std::getline(infile, line))
	{
		//this python looks hairy, not really sure what "name" the elif conditionals refer to...
		//code for reference:
		
		/*if i % 6 == 0:
		 *	name = lines[i].strip("\n").replace(" ","-")
	     *	newPokemon = p.Pokemon(name)
		 *  self.team[name] = newPokemon
		 *elif not (i % 6 == 5):
		 *  self.team[name].addMove(lines[i].replace(" ","").lower().strip("\n"))
		 *elif i == 5:
		 *  self.set_current_out(player_move(True, newPokemon))
		 */
			
	}
}




